### PR TITLE
Feat/melih branch mvp api

### LINKED ISSUE #34: Web API Global Exceptions Handling
### Description:
Implement a global exception handling mechanism for the Web API to ensure that all unhandled exceptions are caught and processed consistently. This will improve the robustness and reliability of the API by providing meaningful error messages and statuses to the clients.

#### Tasks:
1. **Research and Plan**:
   - Investigate best practices for global exception handling in web APIs.
   - Decide on the structure and format of the error responses (e.g., HTTP status codes, error messages, error codes).

2. **Create Custom Exception Classes**:
   - Define custom exception classes for common error scenarios (e.g., `NotFoundException`, `ValidationException`).

3. **Implement Global Exception Handler**:
   - Develop a global exception handler to catch all unhandled exceptions.
   - Ensure the handler logs the exceptions for further analysis and debugging.
   - Map exceptions to appropriate HTTP status codes and error responses.

4. **Middleware Integration**:
   - Integrate the global exception handler into the middleware pipeline to ensure it catches exceptions thrown from any part of the API.

5. **Error Response Format**:
   - Define a consistent error response format.
   - Include relevant details such as error code, message, and any additional information that might help the client understand the error.

 ```json
{
   "success":false,
   "status":"HttpStatusCode.InternalServerError",
   "message":"exception.Message",
   "stackTrace":"exception.StackTrace"
}
```

6. **Logging**:
   - Implement logging for all caught exceptions to facilitate monitoring and debugging.
   - Choose an appropriate logging framework and configure it to log errors consistently.

#### Acceptance Criteria:
- [x] All unhandled exceptions are caught by the global exception handler.
- [x] Error responses follow the defined format and provide meaningful information to clients.
- [x] Exceptions are logged for further analysis and debugging.
- [x] All routes must have this functionality.

### PR DESCRIPTION
**Pull request by Sirmelihy**
FIX #34

### CODE REVIEW

**Review by BharatSoni23 — COMMENTED** [review]
Hello,
My name is Bharat. I'm a dot net developer and I have 2 years of experience. I have checked the code and it's a good approach to handle errors.

**Comment by Sirmelihy** [comment]
> Hello, My name is Bharat. I'm a dot net developer and I have 2 years of experience. I have checked the code and it's a good approach to handle errors.

Hi Bharat,

Thank you for your contribution. May I kindly ask how you discovered our project?

Wishing you the best of luck.

**Comment by BharatSoni23** [comment]
> > Hello, My name is Bharat. I'm a dot net developer and I have 2 years of experience. I have checked the code and it's a good approach to handle errors.
> 
> Hi Bharat,
> 
> Thank you for your contribution. May I kindly ask how you discovered our project?
> 
> Wishing you the best of luck.

Hello @Sirmelihy,
I have searched for a project for contribution and I found your project at the top of my search.
Can you please guide me on how I can contribute to open-source projects?

Thank you

**Comment by fehmisener** [comment]
> > > Hello, My name is Bharat. I'm a dot net developer and I have 2 years of experience. I have checked the code and it's a good approach to handle errors.
> > 
> > 
> > Hi Bharat,
> > Thank you for your contribution. May I kindly ask how you discovered our project?
> > Wishing you the best of luck.
> 
> Hello @Sirmelihy, I have searched for a project for contribution and I found your project at the top of my search. Can you please guide me on how I can contribute to open-source projects?
> 
> Thank you

Hi mate,

Of course you're welcome to contribute our project but it's still in development phase and we usually discuss internally next features and steps. You can watch the repository for the upcoming issues (or you can raise an one) and taking the responsibility of it. Then, all you need to do is following branch creation and clean-code rules.

File: Rater.Business/Configurations/GlobalExceptionHandlingMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            }
+        }
+
+        public async Task HandleExceptionAsync(HttpContext context , Exception ex)
+        {
+            var exceptionType = ex.GetType();
+
+            bool success = false;
+            HttpStatusCode status = HttpStatusCode.InternalServerError;
+            string message = "";
```
Comment (fehmisener) [inline Rater.Business/Configurations/GlobalExceptionHandlingMiddleware.cs]: Why this is `""`and other one is string.Empty

File: Rater.Business/Configurations/GlobalExceptionHandlingMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            string stackTrace = string.Empty;
+
+            if(exceptionType == typeof(UnauthorizedAccessException))
```
Comment (fehmisener) [inline Rater.Business/Configurations/GlobalExceptionHandlingMiddleware.cs]: Do you really need this if structure? Why you don't set `ex.Message`, `ex.StackTrace ?? string.Empty` and `success` above just once. And I feel like this status code extraction should be a new method. Check this implementation:

```csharp
public async Task HandleExceptionAsync(HttpContext context, Exception ex)
{
    var exceptionType = ex.GetType();
    var (success, status) = GetExceptionDetails(exceptionType);
    
    var message = ex.Message;
    var stackTrace = ex.StackTrace ?? string.Empty;

    _logger.LogError(ex, "An error occurred: {Message}, StackTrace: {StackTrace}", message, stackTrace);

    var exceptionResult = JsonSerializer.Serialize(new { success, status, message, stackTrace });
    
    context.Response.ContentType = "application/json";
    context.Response.StatusCode = (int)status;
    await context.Response.WriteAsync(exceptionResult);
}

private static (bool success, HttpStatusCode status) GetExceptionDetails(Type exceptionType)
{
    return exceptionType switch
    {
        Type t when t == typeof(UnauthorizedAccessException) => (false, HttpStatusCode.Unauthorized),
        Type t when t == typeof(InvalidOperationException) => (false, HttpStatusCode.BadRequest),
        Type t when t == typeof(ArgumentException) => (false, HttpStatusCode.BadRequest),
        _ => (false, HttpStatusCode.InternalServerError)
    };
}
```

**Review by fehmisener — CHANGES_REQUESTED** [review]
Check my comments and we are sure that it handles all possible exceptions

**Comment by BharatSoni23** [comment]
> > > > Hello, My name is Bharat. I'm a dot net developer and I have 2 years of experience. I have checked the code and it's a good approach to handle errors.
> > > 
> > > 
> > > Hi Bharat,
> > > Thank you for your contribution. May I kindly ask how you discovered our project?
> > > Wishing you the best of luck.
> > 
> > 
> > Hello @Sirmelihy, I have searched for a project for contribution and I found your project at the top of my search. Can you please guide me on how I can contribute to open-source projects?
> > Thank you
> 
> Hi mate,
> 
> Of course you're welcome to contribute our project but it's still in development phase and we usually discuss internally next features and steps. You can watch the repository for the upcoming issues (or you can raise an one) and taking the responsibility of it. Then, all you need to do is following branch creation and clean-code rules.

Hi @fehmisener 
Thank you for including me to contribute in your project. I have checked your project code and many classes are has unwanted using like 'using System.text' so it will be a great approach to delete them. Can I do this?

**Review by fehmisener — APPROVED** [review]

**Comment by Sirmelihy** [comment]
> > > > > Hello, My name is Bharat. I'm a dot net developer and I have 2 years of experience. I have checked the code and it's a good approach to handle errors.
> > > > 
> > > > 
> > > > Hi Bharat,
> > > > Thank you for your contribution. May I kindly ask how you discovered our project?
> > > > Wishing you the best of luck.
> > > 
> > > 
> > > Hello @Sirmelihy, I have searched for a project for contribution and I found your project at the top of my search. Can you please guide me on how I can contribute to open-source projects?
> > > Thank you
> > 
> > 
> > Hi mate,
> > Of course you're welcome to contribute our project but it's still in development phase and we usually discuss internally next features and steps. You can watch the repository for the upcoming issues (or you can raise an one) and taking the responsibility of it. Then, all you need to do is following branch creation and clean-code rules.
> 
> Hi @fehmisener Thank you for including me to contribute in your project. I have checked your project code and many classes are has unwanted using like 'using System.text' so it will be a great approach to delete them. Can I do this?

Check the issues. I created one , you are welcome to work on it.

**Pull request merged**